Make Pale Moon application specific code buildable

Blocks #22 

Buildability Status:
- [X] Windows
- [X] Linux
- [ ] Mac

```
 3:19.81 /Users/pale/jenkins/workspace/tycho/default/browser/components/shell/nsMacShellService.cpp:35:1: error: static_assert failed "Reference-counted class nsMacShellService should not have a public destructor. Try to make this class's destructor non-public. If that is really not possible, you can whitelist this class by providing a HasDangerousPublicDestructor specialization for it."
 3:19.84 NS_IMPL_ISUPPORTS(nsMacShellService, nsIMacShellService, nsIShellService, nsIWebProgressListener)
 3:19.84 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 3:19.85 ../../../dist/include/nsISupportsImpl.h:1030:3: note: expanded from macro 'NS_IMPL_ISUPPORTS'
 3:19.85   NS_IMPL_ADDREF(aClass)                                                      \
 3:19.85   ^~~~~~~~~~~~~~~~~~~~~~
 3:19.87 ../../../dist/include/nsISupportsImpl.h:564:3: note: expanded from macro 'NS_IMPL_ADDREF'
 3:19.87   MOZ_ASSERT_TYPE_OK_FOR_REFCOUNTING(_class)                                  \
 3:19.87   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 3:19.87 ../../../dist/include/nsISupportsImpl.h:91:3: note: expanded from macro 'MOZ_ASSERT_TYPE_OK_FOR_REFCOUNTING'
 3:19.88   static_assert(!MOZ_IS_DESTRUCTIBLE(X) || \
 3:19.88   ^             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 3:20.44 /Users/pale/jenkins/workspace/tycho/default/browser/components/shell/nsMacShellService.cpp:206:51: error: too few arguments to function call, expected 8, have 7
 3:20.45                       mBackgroundFile, loadContext);
 3:20.45                                                   ^
 3:20.52 ../../../dist/include/nsIWebBrowserPersist.h:86:3: note: 'SaveURI' declared here
 3:20.53   NS_IMETHOD SaveURI(nsIURI *aURI, nsISupports *aCacheKey, nsIURI *aReferrer, uint32_t aReferrerPolicy, nsIInputStream *aPostData, const char * aExtraHeaders, nsISupports *aFile, nsILoadContext *aPrivacyContext) = 0;
 3:20.53   ^
 3:20.53 ../../../dist/include/nscore.h:185:29: note: expanded from macro 'NS_IMETHOD'
 3:20.54 #define NS_IMETHOD          NS_IMETHOD_(nsresult)
 3:20.54                             ^
 3:20.55 ../../../dist/include/nscore.h:123:27: note: expanded from macro 'NS_IMETHOD_'
 3:20.55 #define NS_IMETHOD_(type) virtual type
 3:20.55                           ^
 3:20.60 /Users/pale/jenkins/workspace/tycho/default/browser/components/shell/nsMacShellService.cpp:230:20: error: out-of-line definition of 'OnStatusChange' does not match any declaration in 'nsMacShellService'
 3:20.61 nsMacShellService::OnStatusChange(nsIWebProgress* aWebProgress,
 3:20.61                    ^~~~~~~~~~~~~~
 3:21.17 3 errors generated.
```

That's where I get to for now.

--- a/dom/media/gmp/GMPChild.cpp
+++ b/dom/media/gmp/GMPChild.cpp
@@ -56,9 +56,7 @@ GMPChild::~GMPChild()
 
 static bool
 GetFileBase(const std::string& aPluginPath,
-#if defined(XP_MACOSX)
             nsCOMPtr<nsIFile>& aLibDirectory,
-#endif
             nsCOMPtr<nsIFile>& aFileBase,
             nsAutoString& aBaseName)
 {
@@ -70,11 +68,9 @@ GetFileBase(const std::string& aPluginPath,
     return false;
   }
 
-#if defined(XP_MACOSX)
   if (NS_FAILED(aFileBase->Clone(getter_AddRefs(aLibDirectory)))) {
     return false;
   }
-#endif
 
   nsCOMPtr<nsIFile> parent;
   rv = aFileBase->GetParent(getter_AddRefs(parent));
@@ -95,18 +91,21 @@ GetFileBase(const std::string& aPluginPath,
 }
 
 static bool
+GetFileBase(const std::string& aPluginPath,
+            nsCOMPtr<nsIFile>& aFileBase,
+            nsAutoString& aBaseName)
+{
+  nsCOMPtr<nsIFile> unusedLibDir;
+  return GetFileBase(aPluginPath, unusedLibDir, aFileBase, aBaseName);
+}
+
+static bool
 GetPluginFile(const std::string& aPluginPath,
-#if defined(XP_MACOSX)
               nsCOMPtr<nsIFile>& aLibDirectory,
-#endif
               nsCOMPtr<nsIFile>& aLibFile)
 {
   nsAutoString baseName;
-#ifdef XP_MACOSX
   GetFileBase(aPluginPath, aLibDirectory, aLibFile, baseName);
-#else
-  GetFileBase(aPluginPath, aLibFile, baseName);
-#endif
 
 #if defined(XP_MACOSX)
   nsAutoString binaryName = NS_LITERAL_STRING("lib") + baseName + NS_LITERAL_STRING(".dylib");
@@ -121,7 +120,14 @@ GetPluginFile(const std::string& aPluginPath,
   return true;
 }
 
-#ifdef XP_WIN
+static bool
+GetPluginFile(const std::string& aPluginPath,
+              nsCOMPtr<nsIFile>& aLibFile)
+{
+  nsCOMPtr<nsIFile> unusedlibDir;
+  return GetPluginFile(aPluginPath, unusedlibDir, aLibFile);
+}
+
 static bool
 GetInfoFile(const std::string& aPluginPath,
             nsCOMPtr<nsIFile>& aInfoFile)
@@ -132,9 +138,22 @@ GetInfoFile(const std::string& aPluginPath,
   aInfoFile->AppendRelativePath(infoFileName);
   return true;
 }
-#endif
 
 #if defined(XP_MACOSX) && defined(MOZ_GMP_SANDBOX)
+static nsCString
+GetNativeTarget(nsIFile* aFile)
+{
+  bool isLink;
+  nsCString path;
+  aFile->IsSymlink(&isLink);
+  if (isLink) {
+    aFile->GetNativeTarget(path);
+  } else {
+    aFile->GetNativePath(path);
+  }
+  return path;
+}
+
 static bool
 GetPluginPaths(const std::string& aPluginPath,
                nsCString &aPluginDirectoryPath,
@@ -147,19 +166,8 @@ GetPluginPaths(const std::string& aPluginPath,
 
   // Mac sandbox rules expect paths to actual files and directories -- not
   // soft links.
-  bool isLink;
-  libDirectory->IsSymlink(&isLink);
-  if (isLink) {
-    libDirectory->GetNativeTarget(aPluginDirectoryPath);
-  } else {
-    libDirectory->GetNativePath(aPluginDirectoryPath);
-  }
-  libFile->IsSymlink(&isLink);
-  if (isLink) {
-    libFile->GetNativeTarget(aPluginFilePath);
-  } else {
-    libFile->GetNativePath(aPluginFilePath);
-  }
+  aPluginDirectoryPath = GetNativeTarget(libDirectory);
+  aPluginFilePath = GetNativeTarget(libFile);
 
   return true;
 }
@@ -195,19 +203,10 @@ GetAppPaths(nsCString &aAppPath, nsCString &aAppBinaryPath)
     return false;
   }
 
-  bool isLink;
-  app->IsSymlink(&isLink);
-  if (isLink) {
-    app->GetNativeTarget(aAppPath);
-  } else {
-    app->GetNativePath(aAppPath);
-  }
-  appBinary->IsSymlink(&isLink);
-  if (isLink) {
-    appBinary->GetNativeTarget(aAppBinaryPath);
-  } else {
-    appBinary->GetNativePath(aAppBinaryPath);
-  }
+  // Mac sandbox rules expect paths to actual files and directories -- not
+  // soft links.
+  aAppPath = GetNativeTarget(app);
+  appBinaryPath = GetNativeTarget(appBinary);
 
   return true;
 }
@@ -689,12 +688,7 @@ GetPluginVoucherFile(const std::string& aPluginPath,
                      nsCOMPtr<nsIFile>& aOutVoucherFile)
 {
   nsAutoString baseName;
-#if defined(XP_MACOSX)
-  nsCOMPtr<nsIFile> libDir;
-  GetFileBase(aPluginPath, aOutVoucherFile, libDir, baseName);
-#else
   GetFileBase(aPluginPath, aOutVoucherFile, baseName);
-#endif
   nsAutoString infoFileName = baseName + NS_LITERAL_STRING(".voucher");
   aOutVoucherFile->AppendRelativePath(infoFileName);
   return true;

